linq4j/src/main/java/org/apache/calcite/linq4j/TopNHeap.java

```diff
@@ -0,0 +1,242 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.linq4j;
+
+import org.apache.calcite.linq4j.function.Function1;
+
+import java.util.Comparator;
+
+/**
+ * Implementation of a stable binary heap with a fetch and an offset.
+ * Stable means that if two items are considered equal,
+ * they will appear in the same order as they were offered to the heap.
+ *
+ * @param <TSource> type of the element that will be added to the heap
+ * @param <TKey> type of the key, which the comparator will use for comparisons
+ */
+public class TopNHeap<TSource, TKey> {
```

Comment:
I had the same question with @julianhyde :) Apart from the priority queues I had also in mind a plain `TreeMap`. I did some preliminary benchmarks at some point (WIP in https://github.com/zabetak/jmh-micro-benchmarks/blob/master/core-java/src/main/java/com/github/zabetak/benchmark/PartialSortBenchmark.java) and `TreeMap` was not bad at all. Plus the risk for bugs is much smaller.

Reply:
The advantage of this implementation is that it is stable. If this is not a requirement, it can be replaced by Java's PriorityQueue. `MinMaxPriorityQueue` only uses one array for both heaps.

Reply:
I can't quite see where you are calculating/maintaining the max. That is a simpler problem if you disallow 'offer' after the first 'poll', whereas MinMaxPriorityQueue has to keep two heaps (sharing storage, but still twice the effort). If so, this gives this algorithm an advantage.